File: backend/src/repositories/dynamodb-track.repository.ts
import {
  PutCommand,
  QueryCommand,
  ScanCommand,
  type DynamoDBDocumentClient,
} from '@aws-sdk/lib-dynamodb';

import { BackendEnv } from '../config/env';
import { LoopTrackMetadata } from '../models/track';
import {
  buildTrackLookupKeys,
  buildTrackPrimaryKey,
  buildTracksBySessionKey,
  DYNAMODB_ENTITY_TYPES,
} from './dynamodb-key-design';
import { TrackRepository } from './track.repository';

type DynamoDbTrackItem = LoopTrackMetadata & {
  pk: string;
  sk: string;
  entityType: typeof DYNAMODB_ENTITY_TYPES.track;
  trackId: string;
  gsi2pk: string;
  gsi2sk: string;
};

type DynamoDbTrackRepositoryEnv = Pick<BackendEnv, 'DYNAMODB_METADATA_TABLE_NAME'>;

export class DynamoDbTrackRepository implements TrackRepository {
  constructor(
    private readonly client: DynamoDBDocumentClient,
    private readonly env: DynamoDbTrackRepositoryEnv
  ) {}

  async listTracks() {
    const response = await this.client.send(
      new ScanCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        FilterExpression: 'entityType = :entityType',
        ExpressionAttributeValues: {
          ':entityType': DYNAMODB_ENTITY_TYPES.track,
        },
      })
    );

    return (response.Items ?? [])
      .map((item) => this.toTrack(item as DynamoDbTrackItem))
      .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
  }

  async listTracksBySession(projectId: string, sessionId: string) {
    const keys = buildTracksBySessionKey(projectId, sessionId);

    const response = await this.client.send(
      new QueryCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        KeyConditionExpression: 'pk = :pk AND begins_with(sk, :skPrefix)',
        ExpressionAttributeValues: {
          ':pk': keys.pk,
          ':skPrefix': keys.skPrefix,
        },
      })
    );

    return (response.Items ?? [])
      .map((item) => this.toTrack(item as DynamoDbTrackItem))
      .sort((a, b) => b.updatedAt.localeCompare(a.updatedAt));
  }

  async getTrackById(trackId: string) {
    const lookupKeys = buildTrackLookupKeys(trackId);

    const response = await this.client.send(
      new QueryCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        IndexName: 'gsi2',
        KeyConditionExpression: 'gsi2pk = :gsi2pk AND gsi2sk = :gsi2sk',
        ExpressionAttributeValues: {
          ':gsi2pk': lookupKeys.gsi2pk,
          ':gsi2sk': lookupKeys.gsi2sk,
        },
        Limit: 1,
      })
    );

    const item = response.Items?.[0];

    if (!item) {
      return null;
    }

    return this.toTrack(item as DynamoDbTrackItem);
  }

  async createTrack(track: LoopTrackMetadata) {
    await this.client.send(
      new PutCommand({
        TableName: this.env.DYNAMODB_METADATA_TABLE_NAME,
        Item: this.toItem(track),
        ConditionExpression: 'attribute_not_exists(pk) AND attribute_not_exists(sk)',
      })
    );

    return track;
  }

  async reset() {
    throw new Error('DynamoDB track repository reset is not supported.');
  }

  private toItem(track: LoopTrackMetadata): DynamoDbTrackItem {
    return {
      ...buildTrackPrimaryKey(track.projectId, track.sessionId, track.id),
      ...buildTrackLookupKeys(track.id),
      ...track,
      entityType: DYNAMODB_ENTITY_TYPES.track,
      trackId: track.id,
    };
  }

  private toTrack(item: DynamoDbTrackItem): LoopTrackMetadata {
    return {
      id: item.trackId,
      projectId: item.projectId,
      sessionId: item.sessionId,
      name: item.name,
      durationMs: item.durationMs,
      volume: item.volume,
      isMuted: item.isMuted,
      s3Bucket: item.s3Bucket,
      s3Key: item.s3Key,
      contentType: item.contentType,
      createdAt: item.createdAt,
      updatedAt: item.updatedAt,
    };
  }
}

File: backend-dotnet/src/Loopr.Api/Repositories/DynamoDbTrackRepository.cs
using Amazon.DynamoDBv2.Model;
using Loopr.Api.Configuration;
using Loopr.Api.Domain;
using Microsoft.Extensions.Options;

namespace Loopr.Api.Repositories;

public sealed class DynamoDbTrackRepository(
    IDynamoDbMetadataStore store,
    IOptions<DynamoDbOptions> options
) : ITrackRepository
{
    private const string LookupIndexName = "gsi2";

    private readonly string tableName = options.Value.MetadataTableName;

    public async Task<IReadOnlyList<LoopTrackMetadata>> ListAsync(
        CancellationToken cancellationToken
    )
    {
        var response = await store.ScanAsync(
            new ScanRequest
            {
                TableName = tableName,
                FilterExpression = "entityType = :entityType",
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":entityType"] = DynamoDbAttributeMap.String(DynamoDbEntityTypes.Track),
                },
            },
            cancellationToken
        );

        return (response.Items ?? [])
            .Select(ToTrack)
            .OrderByDescending(track => track.UpdatedAt)
            .ToArray();
    }

    public async Task<LoopTrackMetadata?> GetByIdAsync(
        string trackId,
        CancellationToken cancellationToken
    )
    {
        var lookupKey = DynamoDbMetadataKeys.BuildTrackLookupKey(trackId);
        var response = await store.QueryAsync(
            new QueryRequest
            {
                TableName = tableName,
                IndexName = LookupIndexName,
                KeyConditionExpression = "gsi2pk = :gsi2pk AND gsi2sk = :gsi2sk",
                ExpressionAttributeValues = new Dictionary<string, AttributeValue>
                {
                    [":gsi2pk"] = DynamoDbAttributeMap.String(lookupKey.PartitionKey),
                    [":gsi2sk"] = DynamoDbAttributeMap.String(lookupKey.SortKey),
                },
                Limit = 1,
            },
            cancellationToken
        );

        var item = (response.Items ?? []).FirstOrDefault();

        return item is null ? null : ToTrack(item);
    }

    public async Task<LoopTrackMetadata> CreateAsync(
        LoopTrackMetadata track,
        CancellationToken cancellationToken
    )
    {
        await store.PutItemAsync(
            new PutItemRequest
            {
                TableName = tableName,
                Item = ToItem(track),
                ConditionExpression = "attribute_not_exists(pk) AND attribute_not_exists(sk)",
            },
            cancellationToken
        );

        return track;
    }

    public Task ResetAsync(CancellationToken cancellationToken)
    {
        cancellationToken.ThrowIfCancellationRequested();

        throw new NotSupportedException("DynamoDB track repository reset is not supported.");
    }

    private static Dictionary<string, AttributeValue> ToItem(LoopTrackMetadata track)
    {
        var primaryKey = DynamoDbMetadataKeys.BuildTrackPrimaryKey(
            track.ProjectId,
            track.SessionId,
            track.Id
        );
        var lookupKey = DynamoDbMetadataKeys.BuildTrackLookupKey(track.Id);

        return new Dictionary<string, AttributeValue>
        {
            ["pk"] = DynamoDbAttributeMap.String(primaryKey.PartitionKey),
            ["sk"] = DynamoDbAttributeMap.String(primaryKey.SortKey),
            ["entityType"] = DynamoDbAttributeMap.String(DynamoDbEntityTypes.Track),
            ["projectId"] = DynamoDbAttributeMap.String(track.ProjectId),
            ["sessionId"] = DynamoDbAttributeMap.String(track.SessionId),
            ["trackId"] = DynamoDbAttributeMap.String(track.Id),
            ["name"] = DynamoDbAttributeMap.String(track.Name),
            ["durationMs"] = DynamoDbAttributeMap.Number(track.DurationMs),
            ["volume"] = DynamoDbAttributeMap.Number(track.Volume),
            ["isMuted"] = DynamoDbAttributeMap.Boolean(track.IsMuted),
            ["s3Bucket"] = DynamoDbAttributeMap.String(track.S3Bucket),
            ["s3Key"] = DynamoDbAttributeMap.String(track.S3Key),
            ["contentType"] = DynamoDbAttributeMap.String(track.ContentType),
            ["createdAt"] = DynamoDbAttributeMap.Timestamp(track.CreatedAt),
            ["updatedAt"] = DynamoDbAttributeMap.Timestamp(track.UpdatedAt),
            ["gsi2pk"] = DynamoDbAttributeMap.String(lookupKey.PartitionKey),
            ["gsi2sk"] = DynamoDbAttributeMap.String(lookupKey.SortKey),
        };
    }

    private static LoopTrackMetadata ToTrack(Dictionary<string, AttributeValue> item)
    {
        return new LoopTrackMetadata(
            DynamoDbAttributeMap.ReadString(item, "trackId"),
            DynamoDbAttributeMap.ReadString(item, "projectId"),
            DynamoDbAttributeMap.ReadString(item, "sessionId"),
            DynamoDbAttributeMap.ReadString(item, "name"),
            DynamoDbAttributeMap.ReadInt(item, "durationMs"),
            DynamoDbAttributeMap.ReadDouble(item, "volume"),
            DynamoDbAttributeMap.ReadBoolean(item, "isMuted"),
            DynamoDbAttributeMap.ReadString(item, "s3Bucket"),
            DynamoDbAttributeMap.ReadString(item, "s3Key"),
            DynamoDbAttributeMap.ReadString(item, "contentType"),
            DynamoDbAttributeMap.ReadTimestamp(item, "createdAt"),
            DynamoDbAttributeMap.ReadTimestamp(item, "updatedAt")
        );
    }
}
